examples/functional.rs

```diff
@@ -0,0 +1,34 @@
+use mos6502::cpu;
+use mos6502::memory::Bus;
+use mos6502::memory::Memory;
+use std::fs::read;
+
+fn main() {
+    // Load the binary file from disk
+    let program = match read("examples/asm/functional_test/6502_functional_test.bin") {
+        Ok(data) => data,
+        Err(err) => {
+            println!("Error reading functional test: {}", err);
+            return;
+        }
+    };
+
+    let mut cpu = cpu::CPU::new(Memory::new());
+
+    cpu.memory.set_bytes(0x00, &program);
+    cpu.registers.program_counter = 0x400;
+
+    // run step-by-step
+    let mut old_pc = cpu.registers.program_counter;
+    while cpu.registers.program_counter != 0x3468 {
+        cpu.single_step();
+        println!("{cpu:?}");
+
+        if cpu.registers.program_counter == old_pc {
```

Comment:
To be honest, I blindly took it from here:
https://github.com/marioballano/emudore/blob/5d98f4447409b1d778b9305b0df9aea064d536c5/src/c64.cpp#L134

They are using the same test image. It might be completely wrong, though. The actual repository for the test rom is a bit sparse on details on how to run the test. At least I couldn't find anyting in their `README.md` and the top of the rom file:

https://github.com/Klaus2m5/6502_65C02_functional_tests
https://github.com/Klaus2m5/6502_65C02_functional_tests/blob/master/6502_functional_test.a65

I could not compile the ROM file myself, because I use ca65 and the rom was built with as65. There seems to be a fork of the original repo, but I didn't dig deeper.

Reply:
Update: I compiled the functional test from the sources now. I used https://github.com/amb5l/6502_65C02_functional_tests, which is a port of the original code to ca65, which is the assembler I have on my mac. Now we get even less far in the test. 🤣 

Here's all I get:

```asm
INC $4868
CPU { registers: Registers { accumulator: 0, index_x: 0, index_y: 0, stack_pointer: StackPointer(0), program_counter: 1029, status: Status(PS_NEGATIVE | PS_UNUSED | PS_DISABLE_INTERRUPTS) }
BNE $FFFE
CPU { registers: Registers { accumulator: 0, index_x: 0, index_y: 0, stack_pointer: StackPointer(0), program_counter: 1029, status: Status(PS_NEGATIVE | PS_UNUSED | PS_DISABLE_INTERRUPTS) }
Infinite loop detected!
```

Probably I messed up the compilation or the memory segments.

Reply:
> Update: I compiled the functional test from the sources now. I used https://github.com/amb5l/6502_65C02_functional_tests, which is a port of the original code to ca65, which is the assembler I have on my mac. Now we get even less far in the test. 🤣

Does ca65 have an option to export the labels it knows? Then you could find which infinite loop starts at address 1029 (the reported program counter).

Reply:
Yes, ca65 has a `-g` flag, which adds debug info to the object file. I added that and also added the flags to create a symbols file (`labels.dbg`). I've updated the Rust code to use the labels &mdash; or at least I hope that's how it works.

I'm not getting any label output, just

```
INC $4868
CPU { registers: Registers { accumulator: 0, index_x: 0, index_y: 0, stack_pointer: StackPointer(0), program_counter: 1029, status: Status(PS_NEGATIVE | PS_UNUSED | PS_DISABLE_INTERRUPTS) }
BNE $FFFE
CPU { registers: Registers { accumulator: 0, index_x: 0, index_y: 0, stack_pointer: StackPointer(0), program_counter: 1029, status: Status(PS_NEGATIVE | PS_UNUSED | PS_DISABLE_INTERRUPTS) }
Infinite loop detected!
```

like before. 
I think my setup is wrong and it doesn't even execute anything at the moment.

Reply:
I'm getting label output now after my latest changes and by adjusting the memory segments in the linker config, I get the following output when I run `cargo run --release --example functional`:

```
CLD
TXS
STA $0200 (.test_case)
LDX #$05
BNE $FFF4
DEX
DEX
BEQ $FFDE
BNE $0008
DEY
DEY
BEQ $0017
CMP #$00
LDA #$01
LDA $0200,X (.test_case)
BNE $FFFE
STA $0200 (.test_case)
BNE $0003
CMP #$00
BCC $FFFE
CMP #$01
BCS $FFFE
TAX
BNE $FFFE
BMI $FFFE
BEQ $FFFE
BPL $FFFE
CPY #$00
BCC $FFFE
CPY #$01
BCS $FFFE
LDA $0200,X (.test_case)
BNE $FFFE
STA $0200 (.test_case)
TXS
PHA
PHA
BNE $FFFE
TXA
BNE $FFFE
CMP #$AA
PLA
BNE $FFFE
BNE $FFFE
CPX #$FF
LDA $0200,X (.test_case)
BNE $FFFE
STA $0200 (.test_case)
PHA
BPL $001A (.zpOR)
BCC $001C
BMI $0003
BCS $0003
PHP
CPX #$FE
PLA
BNE $FFFE
CPX #$FF
LDA #$04
PLP
BVS $001B
BEQ $001D
BVC $0003
BNE $000F (.adrl)
PLA
BNE $FFFE
Infinite loop detected!
CPU { registers: Registers { accumulator: 4, index_x: 255, index_y: 0, stack_pointer: StackPointer(255), program_counter: 1336, status: Status(PS_NEGATIVE | PS_DISABLE_INTERRUPTS) }
```